### ISSUE TITLE
False positive for class method without Self annotation 

### ISSUE THREAD

**Issue by tyralla** [python/mypy#15771]
Mypy reports an error for method `save1` (without annotation) but no errors for `save2` and `save3` (with different Self annotations), despite their (almost) identical function bodies:

```python
from copy import deepcopy
from typing import Self

class A:
    states: list[Self]

    def save1(self) -> None:
        self.states.append(deepcopy(self))  # error: Argument 1 to "append" of "list" has incompatible type "A"; expected "Self"  [arg-type]

    def save2(self: Self) -> None:
        self.states.append(deepcopy(self))

    def save3(self) -> Self:
        self.states.append(deepcopy(self))
        return self
```

This bug might be related to #15529 but is not fixed by #15541.
